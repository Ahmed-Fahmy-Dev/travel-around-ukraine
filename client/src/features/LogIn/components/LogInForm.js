import React, { useState } from 'react';
import { Box, Button, Typography, CircularProgress } from '@mui/material';
import { useNavigate } from 'react-router-dom';
import { TextField } from 'formik-mui';
import { Formik, Field, Form } from 'formik';
import * as yup from 'yup';
import postLogIn from '../../../api/postLogIn';

const initialValues = {
  email: '',
  pwd: '',
};

const validationSchema = yup.object().shape({
  email: yup.string().email('It will be an valid email').required('Email is required'),
  pwd: yup.string().min(6, 'Password must have at least 6 characters').required('Password is required!'),
});

const inputBoxSX = (isLoading) => ({
  display: isLoading ? 'none' : 'flex',
  flexDirection: 'column',
  alignItems: 'center',
});

const inputSX = {
  width: { xs: '80%', tablet: '60%', laptop: '40%' },
  my: { xs: '4%', tablet: '3%' },
};

const buttonSX = {
  width: { xs: '86px', mobile: '110px', tablet: '140px', laptop: '150px' },
  px: 0,
  py: { xs: '8px', mobile: '12px', desktop: '14px' },
  mx: 4,
  fontSize: { xs: '12px', laptop: '14px' },
};

const preloaderSX = (isLoading) => ({
  display: isLoading ? 'inline-block' : 'none',
  my: 5,
  width: '80px !important',
  height: '80px !important',
  color: 'secondary',
});

const LogInForm = () => {
  const [errorMassage, setErrorMassage] = useState(null);
  const [isLoading, setIsLoading] = useState(null);

  const navigate = useNavigate();

  const myHandleSubmit = async (values) => {
    setIsLoading(true);

    const userData = {
      loginOrEmail: values.email,
      password: values.pwd,
    };
    const response = await postLogIn(userData).then((res) => res);

    if (response.success) {
      localStorage.setItem('token', response.token);
      setErrorMassage(null);
      navigate('/');
    } else {
      setErrorMassage(response?.password || response?.loginOrEmail);
    }
    setIsLoading(false);
  };

  const handleClickSignUp = () => {
    navigate('/signup');
  };

  return (
    <>
      <CircularProgress sx={preloaderSX(isLoading)} />
      <Formik initialValues={initialValues} validationSchema={validationSchema} onSubmit={myHandleSubmit}>
        {({ handleSubmit, handleChange, values, touched, errors }) => (
          <Box
            component={Form}
            sx={inputBoxSX(isLoading)}
            onSubmit={(e) => {
              e.preventDefault();
              return handleSubmit();
            }}
          >
            <Field
              component={TextField}
              onChange={handleChange}
              value={values.email}
              error={touched.email && Boolean(errors.email)}
              helperText={touched.email && errors.email}
              name="email"
              label="Email"
              sx={inputSX}
            />
            <Field
              component={TextField}
              onChange={handleChange}
              value={values.pwd}
              error={touched.pwd && Boolean(errors.pwd)}
              helperText={touched.pwd && errors.pwd}
              type="password"
              name="pwd"
              label="Password"
              sx={inputSX}
            />
            {errorMassage && <Typography sx={{ color: '#d32f2f', textAlign: 'right' }}>{errorMassage}</Typography>}

            <Box sx={{ my: { xs: 1.5, laptop: 2.5 } }}>
              <Button sx={buttonSX} onClick={handleClickSignUp}>
                sign up
              </Button>
              <Button type="submit" sx={buttonSX}>
                log in
              </Button>
            </Box>
          </Box>
        )}
      </Formik>
    </>
  );
};

export default LogInForm;
