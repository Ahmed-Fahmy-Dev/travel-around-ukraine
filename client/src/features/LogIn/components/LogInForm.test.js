import LogInForm from './LogInForm';
import { BrowserRouter } from 'react-router-dom';
import { render } from '@testing-library/react';

jest.mock('@mui/material/Box', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Button', () => ({ children }) => <button>{children}</button>);
jest.mock('@mui/material/Typography', () => ({ children }) => <p>{children}</p>);
jest.mock('../../../api/postLogIn', () => () => () => {});

// jest.mock('formik-mui/TextField', () => ({ children }) => <input>{children}</input>);

describe('Render LogInForm', () => {
  test('should LogInForm render', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <LogInForm />
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
