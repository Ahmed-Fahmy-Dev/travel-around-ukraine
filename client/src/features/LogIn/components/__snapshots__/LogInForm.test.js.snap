// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render LogInForm should LogInForm render 1`] = `
<DocumentFragment>
  <span
    class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-w61sv4-MuiCircularProgress-root"
    role="progressbar"
    style="width: 40px; height: 40px;"
  >
    <svg
      class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </span>
  <div>
    <div
      class="MuiFormControl-root MuiTextField-root css-1s9g1fu-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":r0:"
        id=":r0:-label"
      >
        Email
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          name="email"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Email
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root css-1s9g1fu-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":r1:"
        id=":r1:-label"
      >
        Password
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r1:"
          name="pwd"
          type="password"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Password
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div>
      <button>
        sing up
      </button>
      <button>
        log in
      </button>
    </div>
  </div>
</DocumentFragment>
`;
